homework-20-12-23/Karine-Vardanyan/create_script.sh

```diff
@@ -0,0 +1,65 @@
+#!/bin/bash
+
+vpc_cidr_block="10.0.0.0/16"
+subnet_cidr_block="10.0.1.0/24"
+region="us-east-1"
+ami_id="ami-xxxxxxxxxxx" # Replace with your desired AMI ID
+instance_type="t2.micro"
+key_pair="machine"
+
+# Create VPC
+vpc_id=$(aws ec2 create-vpc --cidr-block $vpc_cidr_block --query 'Vpc.VpcId' --output text --region $region)
+echo "VPC created with ID: $vpc_id"
+
+# Enable DNS support in the VPC
+aws ec2 modify-vpc-attribute --vpc-id $vpc_id --enable-dns-support
+
+# Create subnet
+subnet_id=$(aws ec2 create-subnet --vpc-id $vpc_id --cidr-block $subnet_cidr_block --availability-zone ${region}a --query 'Subnet.SubnetId' --output text)
+echo "Subnet created with ID: $subnet_id"
+
+# Create Internet Gateway
+internet_gateway_id=$(aws ec2 create-internet-gateway --query 'InternetGateway.InternetGatewayId' --output text)
+echo "Internet Gateway created with ID: $internet_gateway_id"
+
+# Attach Internet Gateway to VPC
+aws ec2 attach-internet-gateway --internet-gateway-id $internet_gateway_id --vpc-id $vpc_id
+
+# Create route table
+route_table_id=$(aws ec2 create-route-table --vpc-id $vpc_id --query 'RouteTable.RouteTableId' --output text)
+echo "Route Table created with ID: $route_table_id"
+
+# Create route to the internet via the Internet Gateway
+aws ec2 create-route --route-table-id $route_table_id --destination-cidr-block 0.0.0.0/0 --gateway-id $internet_gateway_id
+
+# Associate route table with subnet
+aws ec2 associate-route-table --route-table-id $route_table_id --subnet-id $subnet_id
+
+# Create security group
+security_group_id=$(aws ec2 create-security-group --group-name MySecurityGroup --description "My security group" --vpc-id $vpc_id --query 'GroupId' --output text)
+echo "Security Group created with ID: $security_group_id"
+
+# Allow inbound traffic on port 22 (SSH) and outbound traffic to the internet
+aws ec2 authorize-security-group-ingress --group-id $security_group_id --protocol tcp --port 22 --cidr 0.0.0.0/0
+aws ec2 authorize-security-group-ingress --group-id $security_group_id --protocol all --port all --cidr 0.0.0.0/0
+
+# Launch EC2 instance
+instance_id=$(aws ec2 run-instances --image-id $ami_id --instance-type $instance_type --subnet-id $subnet_id --security-group-ids $security_group_id --key-name $key_pair --query 'Instances[0].InstanceId' --output text)
+echo "EC2 instance launched with ID: $instance_id"
+
+# Wait for the instance to be running
+aws ec2 wait instance-running --instance-ids $instance_id --region $region
+
+# Allocate Elastic IP
+allocation_output=$(aws ec2 allocate-address --domain vpc --region $region)
+
+# Extract Allocation ID from the output
+allocation_id=$(echo $allocation_output | jq -r '.AllocationId')
```

Comment:
the parsed data is wrong, so the Elastic IP does not get assigned

Reply:
But I try and my script works for my aws account and Elastic IP is created successfully

Reply:
This is the result after running creation script:

![Screenshot from 2024-01-11 13-26-24](https://github.com/Gev544/devops-hti-2023/assets/47259489/d643c80f-5458-437f-833c-8bd27592ca88)

here you can see the instance is not being assigned an Elastic IP, it is created as you can see in the next screenshot, but it is not being assigned to the instance:

![Screenshot from 2024-01-11 13-27-51](https://github.com/Gev544/devops-hti-2023/assets/47259489/d7b02ddf-5440-4a0e-86b1-164b82d629bd)

![Screenshot from 2024-01-11 13-29-20](https://github.com/Gev544/devops-hti-2023/assets/47259489/750e93d1-3566-4dbb-8bd3-40ea3f0395f3)

After the creation script is complete and I run the cleanup script here's the result:

![Screenshot from 2024-01-11 13-32-11](https://github.com/Gev544/devops-hti-2023/assets/47259489/9316f8b8-8a73-48a0-9adb-d9884e08ee10)

Even though the deletion is working eventually, there are issues when parsing and assigning the Elastic IP to the instance.

Reply:
![screen](https://github.com/Gev544/devops-hti-2023/assets/91372037/b64652c3-fd4e-4504-8231-18c12a6d2463)
![screen2](https://github.com/Gev544/devops-hti-2023/assets/91372037/1a95268f-241d-4c48-8815-12147c24d9a3)
![screen3](https://github.com/Gev544/devops-hti-2023/assets/91372037/48e3acf9-55c8-4b33-b367-a01d1dbd1aef)

Here in these screenshots you can see that the Elastic IP is created and associated with my instance. 
Sorry, how can I see this issue in my terminal to fix this problem?

Reply:
The issue might be we are using different ami-ids for instance creation. I am using **ami-0c7217cdde317cfec**, which is the Free Tier Eligible Ubuntu 22.04 LTS instance. Here's a screenshot.

![Screenshot from 2024-01-11 14-28-16](https://github.com/Gev544/devops-hti-2023/assets/47259489/5abc1f20-26b1-4e10-8837-a0573a3f87f2)

what ami-id are you using?

Reply:
![screen4](https://github.com/Gev544/devops-hti-2023/assets/91372037/43214a9c-868c-4df0-bde8-3e9a057230b4)
No, I use the same ami_id as you ))

Reply:
Can you do a clean run?
Remove all non permanent resources, including Elastic IPs.
Screenshot the whole output and post it here.

Reply:
I incorporated the "usage:permanent" tag logic into the cleanup script. Below, you can find a screenshot indicating the successful execution of the code.
![screen5](https://github.com/Gev544/devops-hti-2023/assets/91372037/083696df-1893-4f3a-81de-1f7fc05123ed)

Reply:
And a clean run of the Creation script would be nice too

Reply:
Running code on different laptops may contain inconsistency of libs/binaries, thats why the same code is working and not working at the same time.
Solution can be by defining all used libs/bins with set versions, or using the containers to avoid environment differences.
@shtajiryan in this case if the code is working on @Karine444's side, just review the code on syntax, using best practices etc...

Reply:
Result of create_script.sh 
![screen8](https://github.com/Gev544/devops-hti-2023/assets/91372037/a8e29bc6-2f79-4fe5-b703-09e6134f1b73)